### PR TITLE
Anopheles refactor part 2 - AnophelesGenomeSequenceData

### PR DESCRIPTION
**Pull request by alimanfoo**
Here we continue working towards #366 by factoring out a class `AnophelesGenomeSequenceData` which assumes responsibility for providing access to a reference genome sequence.

Also here we reorganise the tests to create a new file `conftest.py` which provides test fixtures common to multiple test modules. 

Also here we try out adding coverage reports via codecov.

### CODE REVIEW

File: malariagen_data/af1.py [hunk 1]
```diff
@@ -10,8 +10,8 @@
 
 import malariagen_data  # used for .__version__
 
-MAJOR_VERSION_INT = 1
-MAJOR_VERSION_GCS_STR = "v1.0"
+MAJOR_VERSION_NUMBER = 1
+MAJOR_VERSION_PATH = "v1.0"
```
Comment (alimanfoo) [inline malariagen_data/af1.py:14]: Renamed for consistency with other variable names.

File: malariagen_data/anoph/genome_sequence.py
Comment (alimanfoo) [inline malariagen_data/anoph/genome_sequence.py:1]: This module provides the new AnophelesGenomeSequenceData class.

File: malariagen_data/anopheles.py [hunk 2]
```diff
@@ -750,36 +768,6 @@ def __init__(
...
-    @property
-    @abstractmethod
-    def _genome_ref_id(self):
-        raise NotImplementedError("Must override _genome_ref_id")
-
-    @property
-    @abstractmethod
-    def _genome_ref_name(self):
-        raise NotImplementedError("Must override _genome_ref_name")
-
```
Comment (alimanfoo) [inline malariagen_data/anopheles.py:782]: Refactored to parent class.

File: malariagen_data/anopheles.py [hunk 3]
```diff
@@ -2013,59 +2001,6 @@ def snp_genotypes(
...
-        else:
-            slice_start = None
-        if resolved_region.end:
-            slice_stop = resolved_region.end
-        else:
-            slice_stop = None
-        loc_region = slice(slice_start, slice_stop)
-
-        return d[loc_region]
-
```
Comment (alimanfoo) [inline malariagen_data/anopheles.py:2068]: Refactored to parent class.

File: malariagen_data/util.py [hunk 4]
```diff
@@ -389,14 +389,13 @@ def _prep_geneset_attributes_arg(attributes):
...
-        return Region(feature.contig, int(feature.start), int(feature.end))
-    else:
-        return None
+    if hasattr(resource, "genome_features"):
+        gene_annotation = resource.genome_features(attributes=["ID"])
+        results = gene_annotation.query(f"ID == '{region}'")
+        if not results.empty:
+            # the region is a feature ID
+            feature = results.squeeze()
+            return Region(feature.contig, int(feature.start), int(feature.end))
```
Comment (alimanfoo) [inline malariagen_data/util.py:398]: Small modification here to support testing of the AnophelesGenomeSequenceData class.

File: tests/anoph/conftest.py
Comment (alimanfoo) [inline tests/anoph/conftest.py:1]: This new file defines test fixtures which are common to the base and genome_sequence test modules. Basically, we want a single place to create some test data.

File: tests/anoph/test_base.py
Comment (alimanfoo) [inline tests/anoph/test_base.py:1]: Lots of refactoring here to use pytest fixtures, and to make use of common fixtures defined in conftest.py.

File: poetry.lock
Comment (alimanfoo) [inline poetry.lock:1]: Poetry environment updated to include pytest-cases.

File: tests/anoph/test_base.py [hunk 5]
```diff
@@ -1,221 +1,110 @@
...
+        assert len(df_ss) > 0
+        expected = fixture.release_manifests[release]
+        assert_frame_equal(df_ss[["sample_set", "sample_count"]], expected)
+        assert (df_ss["release"] == release).all()
 
 
-@pytest.mark.parametrize("api", [ag3_public_param, ag3_pre_param, af1_public_param])
-def test_lookup_release(api):
+@parametrize_with_cases("fixture,api", cases=".")
+def test_lookup_release_ag3(fixture, api):
```
Comment (alimanfoo) [inline tests/anoph/test_base.py:107]:
```suggestion
def test_lookup_release(fixture, api):
```

**Review by leehart — APPROVED** [review]

**Comment by alimanfoo** [comment]
Updating the poetry environment has bumped pandas from 1.5 to 2.0, which is then generating some errors around datetime dtype conversions when trying to build the "quarter" column in the sample metadata. I've worked around here by calculating quarter directly.

**Comment by leehart** [comment]
> Updating the poetry environment has bumped pandas from 1.5 to 2.0, which is then generating some errors around datetime dtype conversions when trying to build the "quarter" column in the sample metadata. I've worked around here by calculating quarter directly.

Looks like the tests will need to be updated too, since they use the same method:
```
        # Check that quarter is derived from month, in cases where it is not -1
        df_samples_with_quarter = df_samples.query("quarter != -1")
        df_derived_quarters = (
            df_samples_with_quarter["month"].astype("datetime64[M]").dt.quarter
        )
>       assert np.all(df_samples_with_quarter["quarter"] == df_derived_quarters)
```

**Comment by alimanfoo** [comment]
Thanks Lee :pray:

**Pull request merged**